betty/plugin/config.py

```diff
@@ -0,0 +1,154 @@
+from betty.serde.dump import Dump, minimize, VoidableDump
+from betty.serde.load import Asserter, RequiredField, Assertions, Fields, OptionalField
+
+
+class PluginConfiguration(Configuration):
+    """
+    Configure a single plugin.
+    """
+
+    def __init__(
+        self,
+        plugin_id: str,
+        plugin_label: StaticTranslationLocalizableConfigurationTranslations,
+        *,
+        plugin_description: StaticTranslationLocalizableConfigurationTranslations
+        | None = None,
+    ):
+        super().__init__()
+        self._plugin_id = plugin_id
+        self._plugin_label = StaticTranslationLocalizableConfiguration(plugin_label)
+        self._plugin_description = StaticTranslationLocalizableConfiguration(
+            plugin_description or {}
+        )
+
+    @property
+    def plugin_id(self) -> str:
+        """
+        The configured plugin ID.
+        """
+        return self._plugin_id
+
+    @property
+    def plugin_label(self) -> StaticTranslationLocalizableConfiguration:
+        """
+        The human-readable short plugin label.
+        """
+        return self._plugin_label
+
+    @property
+    def plugin_description(self) -> StaticTranslationLocalizableConfiguration:
+        """
+        The human-readable long plugin description.
+        """
+        return self._plugin_description
+
+    @override
+    def update(self, other: Self) -> None:
+        self._plugin_id = other.plugin_id
+        self._plugin_label = other.plugin_label
+        self._plugin_description = other.plugin_description
+
+    @override
+    @classmethod
+    def load(
+        cls,
+        dump: Dump,
+        configuration: Self | None = None,
+    ) -> Self:
+        asserter = Asserter()
+        plugin_id = asserter.assert_field(
+            RequiredField("plugin_id", Assertions(asserter.assert_str())),
+        )(dump)
+        asserter.assert_record(
+            Fields(
+                RequiredField("plugin_id"),
+                RequiredField(
+                    "plugin_label", Assertions(configuration.plugin_label.load)
+                ),
+                OptionalField(
+                    "plugin_description",
+                    Assertions(configuration.plugin_description.load),
+                ),
+            )
+        )
+        if configuration is None:
+            configuration = cls(plugin_id, Str.plain(""))
+        return configuration
+
+    @override
+    def dump(self) -> VoidableDump:
+        return minimize(
+            {
+                "plugin_id": self.plugin_id,
+                "plugin_label": self.plugin_label.dump(),
+                "plugin_description": self.plugin_description.dump(),
+            }
+        )
+
+    @property
+    def plugin(self) -> type[Plugin]:
+        """
+        The plugin class for this configuration.
+
+        You SHOULD NOT cache the value anywhere, as it *will* change
+        when this configuration changes.
+        """
+        raise AssertionError
```

Comment:
Can this call an event type class factory? Can we then use the same factory to define the two dozen or so hardcoded event types with it?

Reply:
This factory should have a cache so the same definitions result in the same classes, identity-wise.